### PR TITLE
tradingview: init at 2.6.1

### LINKED ISSUE #240783: Package request: TradingView
**Project description**

Desktop application for charting, technical analysis, and trading. It provides traders and investors with a wide range of tools and features to analyze financial markets and make informed trading decisions. TradingView offers real-time and historical market data for various financial instruments such as stocks, cryptocurrencies, forex pairs, commodities, and indices. Users can create interactive charts with a variety of chart types, timeframes, and drawing tools. They can apply technical indicators and overlays to analyze price movements, identify trends, and generate trading signals.

**Metadata**

* homepage URL: https://www.tradingview.com/desktop/
* source URL: https://snapcraft.io/tradingview
* source URL(2): https://aur.archlinux.org/packages/tradingview (Maybe this could help)
* license: Propietary
* platforms: linux

### PR DESCRIPTION
**Pull request by ghost**
## Description of changes

Resolves #240783. I adapted the update script from `spotify`.

~~I've added myself to `maintainer-list.nix` in #260108 (this is why the test is failing; let me know if I should approach this differently).~~

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) (or backporting [23.05 Release notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2305.section.md))
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/applications/finance/tradingview/default.nix [hunk 1]
```diff
@@ -0,0 +1,92 @@
...
+, xorg
+}:
+
+stdenv.mkDerivation rec {
+  pname = "tradingview";
+  version = "2.6.1";
+  revision = "44";
+  src = fetchurl {
+    url = "https://api.snapcraft.io/api/v1/snaps/download/nJdITJ6ZJxdvfu8Ch7n5kH5P99ClzBYV_${revision}.snap";
+    sha512 = "1ddd344d68cfb24774403ce9392c100ae330db49d6e27d71c51913d6b03de69cdb5edc3b5c5c6b25d3169335836ee72ea24536a91d9be6da2f007020c3d134b5";
```
Comment (cafkafk) [inline pkgs/applications/finance/tradingview/default.nix:29]: I converted this to a sri hash like this:

```
nix hash to-sri --type "sha512" 1ddd344d68cfb24774403ce9392c100ae330db49d6e27d71c51913d6b03de69cdb5edc3b5c5c6b25d3169335836ee72ea24536a91d9be6da2f007020c3d134b5 
sha512-Hd00TWjPskd0QDzpOSwQCuMw20nW4n1xxRkT1rA95pzbXtw7XFxrJdMWkzWDbucuokU2qR2b5tovAHAgw9E0tQ==
```

This should allow you to use the more idiomatic:
```suggestion
    hash = "sha512-Hd00TWjPskd0QDzpOSwQCuMw20nW4n1xxRkT1rA95pzbXtw7XFxrJdMWkzWDbucuokU2qR2b5tovAHAgw9E0tQ==";
```

File: pkgs/applications/finance/tradingview/default.nix [hunk 2]
```diff
@@ -0,0 +1,92 @@
...
+
+    addAutoPatchelfSearchPath $out/lib/x86_64-linux-gnu
+    mkdir $out/bin
+    makeWrapper $out/tradingview $out/bin/tradingview --prefix LD_LIBRARY_PATH : ${ lib.makeLibraryPath buildInputs }:$out/lib/x86_64-linux-gnu
+
+    runHook postInstall
+  '';
+
+  meta = with lib; {
+    description = "TradingView desktop platform";
```
Comment (cafkafk) [inline pkgs/applications/finance/tradingview/default.nix:84]: From https://github.com/NixOS/nixpkgs/tree/master/pkgs#meta-attributes

> meta.description must: Not start with the package name.

File: pkgs/applications/finance/tradingview/default.nix [hunk 3]
```diff
@@ -0,0 +1,92 @@
...
+  version = "2.6.1";
+  revision = "44";
+  src = fetchurl {
+    url = "https://api.snapcraft.io/api/v1/snaps/download/nJdITJ6ZJxdvfu8Ch7n5kH5P99ClzBYV_${revision}.snap";
+    sha512 = "1ddd344d68cfb24774403ce9392c100ae330db49d6e27d71c51913d6b03de69cdb5edc3b5c5c6b25d3169335836ee72ea24536a91d9be6da2f007020c3d134b5";
+  };
+
+  nativeBuildInputs = [
+    autoPatchelfHook
+    makeWrapper
```
Comment (cafkafk) [inline pkgs/applications/finance/tradingview/default.nix:34]:
```suggestion
    makeBinaryWrapper
```
This is preffered, as can be seen in https://nixos.org/manual/nixpkgs/stable/#fun-makeWrapper

File: pkgs/top-level/all-packages.nix [hunk 4]
```diff
@@ -42344,4 +42344,6 @@ with pkgs;
   ssl-proxy = callPackage ../tools/networking/ssl-proxy { };
 
   code-maat = callPackage ../development/tools/code-maat {};
+
+  tradingview = callPackage ../development/tools/tradingview {  };
```
Comment (cafkafk) [inline pkgs/top-level/all-packages.nix:42348]: non-blocking: you could consider adding this to `pkgs/by-name instead`

see: https://github.com/NixOS/nixpkgs/blob/master/pkgs/by-name/README.md

**Review by cafkafk — CHANGES_REQUESTED** [review]
It would be cool to have this, thanks for submitting!

**Comment by ProminentRetail** [comment]
> It would be cool to have this, thanks for submitting!

And thanks for the review! :D

**Review by cafkafk — CHANGES_REQUESTED** [review]
Could you perhaps add yourself to the maintainers file in a separate PR, as that should be merged a lot faster? e.g. `maintainers: add prominentretail`, and link to the open PRs you have that depend on it in that PRs description.

**Comment by cafkafk** [comment]
@ofborg build

**Comment by cafkafk** [comment]
@ofborg eval

**Review by cafkafk — COMMENTED** [review]
Sign in is broken. It's very much still usable thou.

```
2023-10-10 14:57:44.621   [DBG] AuthenticationHandler.tryObtainTokenFromClipboard()
2023-10-10 14:57:44.627   [DBG] AuthenticationHandler.acceptRedirectUrl()
2023-10-10 14:57:44.627   [DBG] AuthenticationHandler.getTokenFromRedirectUrl()
2023-10-10 14:57:44.627   [ERR] [AuthenticationHandler.getTokenFromRedirectUrl()] Invalid redirect url gh pr checkout 260133
```

Seems like it is trying (and failing) to get the auth token from the clipboard, `gh pr checkout 260133` is my clipboard contents.

**Comment by ProminentRetail** [comment]
> Sign in is broken. It's very much still usable thou.
> 
> ```
> 2023-10-10 14:57:44.621   [DBG] AuthenticationHandler.tryObtainTokenFromClipboard()
> 2023-10-10 14:57:44.627   [DBG] AuthenticationHandler.acceptRedirectUrl()
> 2023-10-10 14:57:44.627   [DBG] AuthenticationHandler.getTokenFromRedirectUrl()
> 2023-10-10 14:57:44.627   [ERR] [AuthenticationHandler.getTokenFromRedirectUrl()] Invalid redirect url gh pr checkout 260133
> ```
> 
> Seems like it is trying (and failing) to get the auth token from the clipboard, `gh pr checkout 260133` is my clipboard contents.

[It looks like a bug with TradingView](https://www.reddit.com/r/TradingView/comments/1223avp/comment/jdqrltg/?utm_source=share&utm_medium=web2x&context=3). Interestingly, sign in works just fine for me on KDE/Microsoft Edge.

**Review by cafkafk — APPROVED** [review]

**Comment by cafkafk** [comment]
> > Sign in is broken. It's very much still usable thou.
> > ```
> > 2023-10-10 14:57:44.621   [DBG] AuthenticationHandler.tryObtainTokenFromClipboard()
> > 2023-10-10 14:57:44.627   [DBG] AuthenticationHandler.acceptRedirectUrl()
> > 2023-10-10 14:57:44.627   [DBG] AuthenticationHandler.getTokenFromRedirectUrl()
> > 2023-10-10 14:57:44.627   [ERR] [AuthenticationHandler.getTokenFromRedirectUrl()] Invalid redirect url gh pr checkout 260133
> > ```
> > 
> > 
> > Seems like it is trying (and failing) to get the auth token from the clipboard, `gh pr checkout 260133` is my clipboard contents.
> 
> [It looks like a bug with TradingView](https://www.reddit.com/r/TradingView/comments/1223avp/comment/jdqrltg/?utm_source=share&utm_medium=web2x&context=3). Interestingly, sign in works just fine for me on KDE/Microsoft Edge.
> 
> **EDIT:** [This](https://www.reddit.com/r/TradingView/comments/1223avp/comment/jkcbhdw/?utm_source=share&utm_medium=web2x&context=3) comment is interesting: does it work with a different browser? Are you using GNOME/Wayland?

Also yea, scrolled down and copied it, and that seemed to work.

**Pull request merged**